Updated the `osgearth` recipe to support v3.7.2.

tylerbrawl opened the pull request:
### Summary
Changes to recipe:  **osgearth/3.7.2**

#### Motivation
The official Conan recipe for `osgearth` (located at `recipes\osgearth\all\conanfile.py`) appears to have been neglected for a long time. Consider, for example, the following:

  * There existed several dependency conflicts based on the versioned references specified in the `osgearth` recipe.

  * OpenSceneGraph and GDAL were not building on the latest version of the MSVC. (These are both mandatory dependencies of `osgearth`.)

  * Several of the member functions of `OsgearthConan` made use of old Conan 1 facilities, such as the old `conans.CMake` type, which are either deprecated or removed in Conan 2.

#### Details
This PR updates the `OsgearthConan` class (that is, the official Conan recipe for `osgearth`) to support Conan 2 in addition to a new version of the library - v3.7.2. (Support for both v3.2 and v3.3 has been retained in this commit. However, this is subject to change pending a code review.) This required a significant number of changes to be made to its implementation. Along the way, several bugs in the previous implementation were encountered and fixed. These are as follows:

  * The `OsgearthConan::generate()` member function will now set the `CMAKE_*_POSTFIX` CMake variables in the CMake cache in order to match what is expected by osgDB. Specifically, any plugins which are loaded by osgDB in a shared/DLL build of OpenSceneGraph are expected to have a postfix of `OSG_LIBRARY_POSTFIX` in their file name. This is a preprocessor macro which is set to match the value of `CMAKE_*_POSTFIX` during the build of osgDB; since OpenSceneGraph uses a particular set of values for the `CMAKE_*_POSTFIX` variables, we want the osgDB plugins generated by `osgearth` to also use these values.

  * For some reason, the `OsgearthConan::requirements()` member function would add `zstd` as a dependency when the `build_zip_plugin` option was set to `True`. In reality, however, the `osgdb_zip` CMake target relies on `libzip`, and not `zstd`. This dependency has thus been adjusted accordingly.

  * The official Conan recipe for `rocksdb` will delete the C++ headers from the packaged include directory. (A comment claims that this is done for the sake of ABI stability.) However, `osgearth` relies on these headers to exist in order to build the `osgdb_osgearth_cache_rocksdb` CMake target. This means that we cannot use a shared build of `rocksdb` in order to build that particular plugin. Thus, the `OsgearthConan::validate()` member function has been updated to check for this particular case.

There are some additional caveats which I would like to mention here. `osgearth` includes an optional `imgui` integration/nodekit which has historically been very problematic. Consider, for example, the confusing evolution of how this integration was distributed:

  * `osgearth` v3.2 includes an embedded version of `imgui` (and `imnodes`) which is used to build the `imgui` nodekit, but only if GLEW exists as a dependency.

  * `osgearth` v3.3 removes the embedded version of `imgui` (and `imnodes`), effectively rendering the `imgui` nodekit unusable (that is, unless the user actually copies their own version of `imgui` into the source directory).

  * `osgearth` v3.7.2 yet again includes an embedded version of `imgui` (and `imnodes`), and yet again, GLEW must exist as a dependency in order to build the `imgui` nodekit. This time, however, there is an additional requirement: the `OSGEARTH_BUILD_IMGUI_NODEKIT` CMake variable must also be set.

It should be possible to patch `osgearth` to ensure consistency between all of our supported versions. However, this is expected to take a significant amount of effort. This PR thus always disables the `imgui` nodekit. Future work would be to add optional support for it in the form of a proposed `build_imgui_nodekit` option.

There also appears to be some issues with the `build_procedural_nodekit` option in older versions of `osgearth`. Specifically, the `osgEarthProcedural` CMake target does not seem to build on the MSVC for these versions. The `OsgearthConan::config_options()` member function seems to check for this, but doing so there is arguably pointless because any changes made to that option from within `OsgearthConan::config_options()` will be overwritten by the value specified in, e.g., the current Conan profile. No explanation is given as to why this check is done, but it seems incorrect. (It should be noted that this does not appear to be an issue for `osgearth` v3.7.2; the `osgEarthProcedural` CMake target builds fine on my machine with the MSVC in that version.)

fixes #25927
closes https://github.com/conan-io/conan-center-index/issues/29027

### Blocking PRs
The following PRs contain changes which must be merged before the changes in this PR can have any meaningful impact:

* [ ] #28121
* [ ] #28126


---
- [X] Read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md)
- [X] Checked that this PR is not a duplicate: [list of PRs by recipe](https://github.com/conan-io/conan-center-index/discussions/24240)
- [x] Tested locally with at least one configuration using a recent version of Conan

tylerbrawl committed:
Updated the `osgearth` recipe to support v3.7.2.

The official Conan recipe for `osgearth` (located at
`recipes\osgearth\all\conanfile.py`) appears to have been neglected
for a long time. Consider, for example, the following:

  * There existed several dependency conflicts based on the
    versioned references specified in the `osgearth` recipe.

  * OpenSceneGraph and GDAL were not building on the latest version
    of the MSVC. (These are both mandatory dependencies of
    `osgearth`.)

  * Several of the member functions of `OsgearthConan` made use of
    old Conan 1 facilities, such as the old `conans.CMake` type,
    which are either deprecated or removed in Conan 2.

This commit updates the `OsgearthConan` class (that is, the
official Conan recipe for `osgearth`) to support Conan 2 in
addition to a new version of the library - v3.7.2. (Support for
both v3.2 and v3.3 has been retained in this commit. However, this
is subject to change pending a code review.) This required a
significant number of changes to be made to its implementation.
Along the way, several bugs in the previous implementation were
encountered and fixed. These are as follows:

  * The `OsgearthConan::generate()` member function will now set
    the `CMAKE_*_POSTFIX` CMake variables in the CMake cache in
    order to match what is expected by osgDB. Specifically, any
    plugins which are loaded by osgDB in a shared/DLL build of
    OpenSceneGraph are expected to have a postfix of
    `OSG_LIBRARY_POSTFIX` in their file name. This is a
    preprocessor macro which is set to match the value of
    `CMAKE_*_POSTFIX` during the build of osgDB; since
    OpenSceneGraph uses a particular set of values for the
    `CMAKE_*_POSTFIX` variables, we want the osgDB plugins
    generated by `osgearth` to also use these values.

  * For some reason, the `OsgearthConan::requirements()` member
    function would add `zstd` as a dependency when the
    `build_zip_plugin` option was set to `True`. In reality,
    however, the `osgdb_zip` CMake target relies on `libzip`, and
    not `zstd`. This dependency has thus been adjusted accordingly
    in this commit.

  * The official Conan recipe for `rocksdb` will delete the C++
    headers from the packaged include directory. (A comment claims
    that this is done for the sake of ABI stability.) However,
    `osgearth` relies on these headers to exist in order to build
    the `osgdb_osgearth_cache_rocksdb` CMake target. This means
    that we cannot use a shared build of `rocksdb` in order to
    build that particular plugin. Thus, this commit updates the
    `OsgearthConan::validate()` member function to check for this
    particular case.

There are some additional caveats which I would like to mention
here. `osgearth` includes an optional `imgui` integration/nodekit
which has historically been very problematic. Consider, for
example, the confusing evolution of how this integration was
distributed:

  * `osgearth` v3.2 includes an embedded version of `imgui` (and
    `imnodes`) which is used to build the `imgui` nodekit, but only
    if GLEW exists as a dependency.

  * `osgearth` v3.3 removes the embedded version of `imgui` (and
    `imnodes`), effectively rendering the `imgui` nodekit unusable
    (that is, unless the user actually copies their own version of
    `imgui` into the source directory).

  * `osgearth` v3.7.2 yet again includes an embedded version of
    `imgui` (and `imnodes`), and yet again, GLEW must exist as a
    dependency in order to build the `imgui` nodekit. This time,
    however, there is an additional requirement: the
    `OSGEARTH_BUILD_IMGUI_NODEKIT` CMake variable must also be set.

It should be possible to patch `osgearth` to ensure consistency
between all of our supported versions. However, this is expected to
take a significant amount of effort. This commit thus always
disables the `imgui` nodekit. Future work would be to add optional
support for it in the form of a proposed `build_imgui_nodekit`
option.

There also appears to be some issues with the
`build_procedural_nodekit` option in older versions of `osgearth`.
Specifically, the `osgEarthProcedural` CMake target does not seem
to build on the MSVC for these versions. The
`OsgearthConan::config_options()` member function seems to check
for this, but doing so there is arguably pointless because any
changes made to that option from within
`OsgearthConan::config_options()` will be overwritten by the value
specified in, e.g., the current Conan profile. No explanation is
given as to why this check is done, but it seems incorrect. (It
should be noted that this does not appear to be an issue for
`osgearth` v3.7.2; the `osgEarthProcedural` CMake target builds
fine on my machine with the MSVC in that version.)

Fixes: #25927

tylerbrawl commented:
I am marking this PR as a draft because I have yet to update the `recipes\osgearth\all\test_package\conanfile.py` testing recipe. However, I have tested the new `osgearth` recipe locally on my own machine under several different configurations.

tylerbrawl commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -1,9 +1,11 @@
 import os
-import functools
 
 required_conan_version = ">=1.33.0"
```
This is almost certainly wrong. However, I am not sure what to put here. The [Consuming Recipes documentation](https://github.com/conan-io/conan-center-index/blob/master/docs/consuming_recipes.md) suggests that we should be supporting Conan 2.0 (or, to be more precise, Conan 2.x, which I interpret as meaning "at least Conan 2.0"). Yet, I have seen several recipes in this repository whose `required_conan_version` value claims to still support some version of Conan 1.

tylerbrawl commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -22,6 +24,33 @@ class OsgearthConan(ConanFile):
+        # "build_imgui_nodekit": [True, False],
+
         "build_procedural_nodekit": [True, False],
         # "build_triton_nodekit": [True, False],
```
The following options have historically been commented out in this recipe without explanation:

* `build_triton_nodekit`
* `build_silverlining_nodekit`
* `with_basisu`

(NOTE: `with_glew` also existed as a commented-out option prior to this commit. However, it was removed because a `build_imgui_nodekit` option would be better suited in practice, since `osgearth` only really ever used GLEW to support its embedded versions of `imgui`.)

I have not tried re-enabling support for these options in this PR. However, it would be helpful to understand _why_ these options aren't (or perhaps weren't) supported.

tylerbrawl replied:
Thanks for clarifying! I do have another question, though. Is it better to leave these so-called "hypothetical options" as commented-out code in the recipe, or should we just be removing any and all mentions of them?

AbrilRBS replied:
Let's remove them althogether, and users that might have a need for them will know to ask for the feature itself :)

tylerbrawl commented on recipes/osgearth/all/conandata.yml:
```diff
@@ -1,11 +1,17 @@
     url: https://github.com/gwaldron/osgearth/archive/refs/tags/osgearth-3.2.tar.gz
 patches:
+  '3.7.2':
+    - patch_file: patches/3.7.2-osgearth.patch
```
The [`conandata.yml` documentation](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/conandata_yml_format.md#patches-fields) would indicate that there are numerous problems with this new patch:

* There is no `patch_description` for this patch.
* There are no `patch_type` or `patch_source` fields for this patch. _(NOTE: These are only listed as recommended requirements.)_

I left this patch undocumented because it was created by migrating the changes from the other patches to v3.7.2, and said patches were themselves undocumented. Thus, I couldn't really tell you why (or even if, honestly) this patch needs to exist without further investigation.

That begs the question, though: Why aren't the other patches documented?

tylerbrawl replied:
The previous patch files were removed in commit 741efa258355c4cf8c971fe861076572a6818f6b. However, the `3.7.2-osgearth.patch` file remains undocumented for the reasons given above.

tylerbrawl commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -65,150 +93,202 @@ class OsgearthConan(ConanFile):
         if self.options.with_geos:
             self.requires("geos/3.11.1")
         if self.options.with_sqlite3:
             self.requires("sqlite3/[>=3.42 <4]")
```
The [Dependencies documentation](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/dependencies.md#version-ranges) would indicate that this version range specification for `sqlite3` is disallowed. Yet, it predates the changes in this PR...

tylerbrawl replied:
Okay, then I'll go ahead and mark this comment as resolved. Thanks!

tylerbrawl commented on recipes/openscenegraph/all/conandata.yml:
```diff
@@ -36,3 +36,13 @@ patches:
       patch_source:
         - https://github.com/openscenegraph/OpenSceneGraph/pull/1296
         - https://github.com/openscenegraph/OpenSceneGraph/pull/951
+    - patch_file: patches/fix-MSVC-_FPOSOFF-removal.patch
```
The existing patches for `openscenegraph` are numbered (that is, they have a file name of the form `XXXX-YY.patch`, where `XXXX` is some number). However, I left the new patches in this PR unnumbered because of the fact that entries in the range [2, 4] are missing. Thus, there is some ambiguity regarding which numbers should be assigned to the new patches:

* `0012-fix-MSVC-_FPOSOFF-removal.patch` and `0013-fix-MSVC-unnamed-typedef-struct.diff` arise from starting after the last numbered patch, `0011-remove-deprecated-register.patch`.
* `0002-fix-MSVC-_FPOSOFF-removal.patch` and `0003-fix-MSVC-unnamed-typedef-struct.diff` arise from attempting to fill in the "gap" between `0001-fix-to_cmake_path-usage.patch` and `0005-use-JPEG-target-for-plugin.patch`.

What is the suggestion here?

AbrilRBS commented:
Hi @tylerbrawl thanks a lot for taking the time create a PR for this, we truly appreciate it!

Just skimmed the PR for now, but:

> Support for both v3.2 and v3.3 has been retained in this commit. However, this is subject to change pending a code review.

If this simplifies the recipe, we'd be happy to drop support for the older versions, and keep only the latest one, thanks!

tylerbrawl commented:
> If this simplifies the recipe, we'd be happy to drop support for the older versions, and keep only the latest one, thanks!

That would certainly simplify the recipe, but the [Sources and Patches documentation](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/sources_and_patches.md#removing-old-versions) states that we should include support for at least two versions of the "latest major release." (I assume that would be `osgearth` v3.x.y in this case.) Would I also need to add support for `osgearth` v3.7.1?

tylerbrawl committed:
Updated the `test_package` recipe for `osgearth`.

This commit updates the testing recipe for `osgearth`. This recipe
is located at `recipes\osgearth\all\test_package\conanfile.py`.
The root `CMakeLists.txt` for this testing recipe has been updated
to resemble the examples given in
[the Conan 2 documentation](https://docs.conan.io/2/tutorial/creating_packages/test_conan_packages.html),
as has the `TestPackageConan` class defined in
`recipes\osgearth\all\test_package\conanfile.py`.

During this testing, it was revealed that the `test_package` CMake
target managed by the testing recipe failed to build. This was
ultimately because Conan does not make the headers of dependencies
transitive by default; this must be explicitly requested when
calling the `ConanFile::requires()` member function by setting the
`transitive_headers` parameter to `True`. Since the public headers
of `osgearth` will frequently attempt to `#include` the public
headers of `openscenegraph`, it makes sense to use transitive
headers for that particular dependency. This conclusion resulted in
the `OsgearthConan::requirements()` member function being updated
accordingly in this commit.

tylerbrawl commented:
I have updated the `test_package` recipe for `osgearth` in commit 5ef389410b97a396da280ccf714e91676f735001. I will mark this PR as ready for review.

tylerbrawl committed:
Removed support for older versions of `osgearth`.

This commit removes support for the following versions of
`osgearth` from its official Conan recipe:

  * v3.2
  * v3.3

This change came from a suggestion by @AbrilRBS that the older
versions could be removed if doing so would simplify the final
`osgearth` recipe. As can be seen from the changes in this commit,
this did, in fact, result in a much simpler implementation for the
`OsgearthConan` class. The `CMakeLists.txt` file which was
previously being injected into the `osgearth` build (via
`OsgearthConan.exports_source`) has been removed in this commit, as
the CMake scripts used in `osgearth` v3.7.2 make proper use of the
`find_package()` CMake command (in most cases, anyways).

This commit also deprecates the following options, as described in
[the FAQs documentation](https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#can-i-remove-an-option-from-a-recipe):

  * `build_leveldb_cache`
  * `with_draco`

These options are ultimately useless in `osgearth` v3.7.2. In
previous versions, they were used to set the values of various
CMake variables, but said variables no longer exist.

tylerbrawl commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -65,150 +93,202 @@ class OsgearthConan(ConanFile):
-    def _build_subfolder(self):
-        return "build_subfolder"
+    def _minimum_cpp_standard(self):
+        return (11 if Version(self.version) < "3.7.2" else 14)
```
This is no longer relevant now that support for older versions of `osgearth` has been removed. However, it may still help to update the documentation to address questions like this one.

AbrilRBS commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -22,18 +23,37 @@ class OsgearthConan(ConanFile):
         # "build_triton_nodekit": [True, False],
         # "build_silverlining_nodekit": [True, False],
-        "build_leveldb_cache": [True, False],
+        "build_leveldb_cache": [True, False, "deprecated"],
```
Let's remove them! As this recipe has not been yet ported to Conan 2, there are no preexisting users that would break here, so removing these options althogether is the way to go too :)

AbrilRBS commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -65,150 +93,202 @@ class OsgearthConan(ConanFile):
-            tools.check_min_cppstd(self, 11)
-        elif self.settings.compiler == "apple-clang":
-            raise ConanInvalidConfiguration("With apple-clang cppstd needs to be set, since default is not at least c++11.")
+            check_min_cppstd(self, self._minimum_cpp_standard)
```
We usually remove such `set(CMAKE_CXX_STANDARD ...` calls, to let user profiles choose the final version, but lately we're looking into just keeping whatever upstream does, but for a final answer on this, I'll ping @jcar87

AbrilRBS commented on recipes/gdal/pre_3.5.0/conanfile.py:
```diff
@@ -172,6 +172,9 @@ def export_sources(self):
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
+        # gdal doesn't build in C++20 mode with latest MSVC, clamp to 17.
+        if is_msvc(self) and int(self.settings.get_safe("compiler.cppstd")) > 17:
+            self.settings.compiler.cppstd = 17
```
But this probably should go into a separate PR so that the focus of this PR remains on osgearth, yes :)

tylerbrawl replied:
Thanks for pointing this out. I may or may not have copy-pasted this from a recipe used internally at my work...

AbrilRBS commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -64,151 +84,179 @@ class OsgearthConan(ConanFile):
+        cmake.install()
 
         if self.options.install_shaders:
-            rename(self, os.path.join(self.package_folder, "resources"), os.path.join(self.package_folder, "res"))
+            rename(self, os.path.join(self.package_folder, "share"), os.path.join(self.package_folder, "res"))
```
We might want to rethink this rename. If the library is installing the shaders in this folder, is there a reason for us to move them somewhere else?

tylerbrawl replied:
My initial thoughts were that we wanted to keep the shader installation directory the same between `osgearth` versions. In other words, if the user built, e.g., `osgearth` v3.3 with the `install_shaders` option set, then the resulting directory would be the same if they decided to update to `osgearth` v3.7.2.

Now that we have removed support for the older `osgearth` versions, though, that may not be necessary. Is it okay for us to do that, or should we be maintaining compatibility with the old recipe as closely as possible?

AbrilRBS replied:
> or should we be maintaining compatibility with the old recipe as closely as possible?

In this case, no, because Conan 2 users are to-date not able to build this recipe, so there's no need to keep compatibility with those

AbrilRBS reviewed (commented):
Some comments on the osgearth recipe itself, which looks great so far, thanks a lot for taking the time to contribute it, we appreciate it

As for gdal and openscenegraph, different PRs for each would ease their review by quite a lot :)

tylerbrawl commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -22,6 +24,33 @@ class OsgearthConan(ConanFile):
+        #
+        # For now, we simply disable the imgui nodekit.
+        #
+        # "build_imgui_nodekit": [True, False],
```
I guess @AbrilRBS answered my question in [this comment](https://github.com/conan-io/conan-center-index/pull/28101#discussion_r2258026584). I have no immediate use for it, so I'll leave it out for this PR.

tylerbrawl replied:
All of the comments referencing a `build_imgui_nodekit` option were removed from the `osgearth` recipe in 07b0a786026f5b6d77cafaacf3a4d4bd55304190.

tylerbrawl commented:
> As for gdal and openscenegraph, different PRs for each would ease their review by quite a lot :)

Sure! For OSG in particular, one of the patches already exists in a separate open PR as #26231. I could create a new PR containing both the duplicate patch and the new patch if we want to just close that one. Otherwise, I can wait for that one to be handled.

tylerbrawl committed:
Added version ranges for some dependencies.

Co-authored-by: Abril Rincón Blanco <5364255+AbrilRBS@users.noreply.github.com>

tylerbrawl committed:
Simplified the `OsgearthConan::build()` member function.

Apparently, the source files can be patched directly within
`ConanFile::source()`. Thanks to @AbrilRBS for the suggestion!

tylerbrawl committed:
Removed several unused and commented-out options from the `osgearth` recipe.

This commit simplifies the `osgearth` recipe by removing a number
of options which were ultimately left unused. It also removes
several "hypothetical options" - that is, comments which usually
take the form of code declaring or accessing an option. The
removals can be categorized into one of several groups.

First, the following deprecated options were removed in this
commit:

  * `build_leveldb_cache`
  * `with_draco`

These options have no use whatsoever in `osgearth` v3.7.2. They
were first converted into deprecated options based on the general
guidance provided in
[the FAQs](https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#can-i-remove-an-option-from-a-recipe).
During a code review, however, @AbrilRBS suggested that they could
simply be removed because the `osgearth` recipe had not been
updated to support Conan 2 prior to this commit.

Second, the following so-called "hypothetical options" were
removed, as suggested during a code review:

  * `build_imgui_nodekit`
  * `build_silverlining_nodekit`
  * `build_triton_nodekit`
  * `with_basisu`

Finally, several additional options were removed in this commit for
ultimately being pointless:

  * `enable_nvtt_cpu_mipmaps`: In `osgearth` v3.3, this option
    would be used to set the `OSGEARTH_ENABLE_NVTT_CPU_MIPMAPS`
    CMake variable. However, this variable is no longer used in
    `osgearth` v3.7.2.

  * `with_sqlite3`: In `osgearth` v3.7.2, `sqlite3` is a mandatory
    dependency. Thus, it does not make much sense to retain this
    option.

tylerbrawl committed:
Modified the shader installation directory for `osgearth`.

The official Conan recipe for `osgearth` would historically install
shaders into a directory called `res`, located at the root of the
package folder. During a code review, however, @AbrilRBS asked that
the shader installation directory used by `osgearth` be retained,
rather than renamed to match the behavior of previous versions of
the `osgearth` recipe. This is because the old recipe had not been
updated to Conan 2 prior to #28101.

Based on that feedback, this commit updates the
`OsgearthConan::package()` member function to no longer attempt to
rename the shader installation directory. For all intents and
purposes, future recipe versions should consider the `share` folder
as the destination for installed shaders.

AbrilRBS commented:
@tylerbrawl ping me once you remove gdal and openscenegraph from this PR, and I'll trigger the CI to build this PR :)

tylerbrawl committed:
Removed the `openscenegraph` changes from #28101.

This commit removes the changes to the `openscenegraph` recipe
which were made pursuant to #28101. Instead, these changes are now
contained within PR #28121.

tylerbrawl committed:
Removed the `gdal` changes from #28101.

This commit removes the changes to the `gdal` recipe which were
made pursuant to #28101. Instead, these changes are now contained
within PR #28126.

tylerbrawl commented:
@AbrilRBS - I have created #28121 and #28126 for `openscenegraph` and `gdal`, respectively. Only the `osgearth` recipe has been modified in this PR.

tylerbrawl committed:
Fixed several issues reported by the Recipe Linter.

This commit addresses several issues which the Recipe Linter
diagnosed for the `osgearth` recipe. These mostly consist of unused
`import` statements, but it did warn about a missing `package_type`
attribute for `OsgearthConan`. (There was also an "error" for not
using `conan.tools.files.export_conandata_patches()`, although I
suspect that the previous use of `OsgearthConan.exports_sources`
achieved equivalent results.)

This commit also updates the `required_conan_version` value for the
`osgearth` recipe to `">=2.0"` based on code review feedback. It
still isn't quite clear if this is correct, though, because I have
not actually tested the recipe at that version.

AbrilRBS committed:
Fix gdal, cleanup

AbrilRBS commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -23,192 +25,166 @@ class OsgearthConan(ConanFile):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+        apply_conandata_patches(self)
+
+    def _get_library_postfix(self, build_type: str) -> str:
```
Hi @tylerbrawl - coming back to this, the only remaining doubt about this recipe for me now is this function.

Could you share some more info about this? Correct me if I'm missing something, but the idea is that the dependency openscenegraph expects to load plugins with a given postfix when using it, with no way to override the behaviour, and that we then need to create our libraries with said postfixes?

I wanted to know more because we usually try to refrain from modifying the library names that recipes generate unless absolutely necessary, so I'm wondering if there could be simpler/different approaches :)

AbrilRBS replied:
Yes, I think raising this with upstream would be worth it if only to let them know.

Do note that other package managers that have this library seem not to care about the output name of the plugins, so I wonder then how they are making them work, if at all

sebhmg replied:
@AbrilRBS , see how this logic was already present in the original recipe:
https://github.com/conan-io/conan-center-index/pull/28101/files#diff-33d8c7904e00023d56217e34e179f3e7aeaf424518888d0259fbf3e4089770bbL204

It is possible that other plugins do not have it in place and the problem went unnoticed, as they would still load fine when their build_type is `release`.

uilianries commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -23,192 +25,166 @@ class OsgearthConan(ConanFile):
-    @property
-    def _build_subfolder(self):
-        return "build_subfolder"
+        # The official Conan recipe for rocksdb will delete all of the C++ headers from the packaged include directory
```
```suggestion
```

The PR #28425 fixed it.

uilianries commented on recipes/osgearth/all/conanfile.py:
```diff
@@ -23,192 +25,166 @@ class OsgearthConan(ConanFile):
-        if self.options.build_leveldb_cache:
-            self.requires("leveldb/1.22")
         if self.options.build_rocksdb_cache:
             self.requires("rocksdb/6.29.5")
```
```suggestion
            self.requires("rocksdb/10.5.1")
```

It would be great checking it if work with RockSDB 10.5.1 due the fix for headers

uilianries reviewed (commented):
Partial review. 

@tylerbrawl Thank you for your PR!

sebhmg committed:
use newer rocksDB version, which fixed its recipe for C++ headers

sebhmg committed:
PATH for bin and LD_LIBRARY_PATH are managed automatically for VirtualRunEnv

sebhmg commented:
> As the changes you propose are made on top of this PR, the better approach would be to create a PR that targets the base branch for this PR inside the contributor's fork,

thanks @AbrilRBS
I had tried to create PR targeting [tylerbrawl:tnichols/features/osgearth-3.7.2](https://github.com/tylerbrawl/conan-center-index/tree/tnichols/features/osgearth-3.7.2), but I could not. It seems the tylebrawl fork does not allow me to:
<img width="266" height="150" alt="image" src="https://github.com/user-attachments/assets/572a9945-c7e9-4c29-a5a6-c73c154dc206" />

sebhmg commented:
> Do note though that the showstopper for this PR for now is understanding https://github.com/conan-io/conan-center-index/pull/28101/files#diff-33d8c7904e00023d56217e34e179f3e7aeaf424518888d0259fbf3e4089770bbR163, which we discussed in [#28101 (comment)](https://github.com/conan-io/conan-center-index/pull/28101#discussion_r2290742450), and we would appreciate if you had any usage insights for those plugins, as we are doubtful if this would be the best approach for them

@AbrilRBS , I do not know too much about the mechanism of OSG to load plugins, but it seems this logic in the recipe was there before. It is likely still correct. I have replied in https://github.com/conan-io/conan-center-index/pull/28101#discussion_r2290742450

AbrilRBS committed:
OsgEarth is a C++17 library

AbrilRBS committed:
Disable C++17 hardcode

AbrilRBS committed:
Listen to linter

AbrilRBS committed:
Fix pattern

AbrilRBS commented:
Working on getting this recipe ready, will push some changes shortly once the last issues with the recipe are ironed out

jcar87 commented:
I believe this was superseded by the now-merged https://github.com/conan-io/conan-center-index/pull/29755 - please respond to this comment if there's anything missing or left to do, thanks!
